### PR TITLE
fix Array#sample randomness

### PR DESCRIPTION
**Pull request by tompng**
Fixes issue #3723

<hr>
1. Is this pull-request complete?

  - [x] Yes, this pull-request is ready to be reviewed and merged.
  - [ ] No, this pull-request is a work-in-progress.

2. Does this pull request fix an issue with an existing feature or introduce a new feature?

  - [x] It fixes an issue with an existing features.
  - [ ] It introduces a new feature.

3. Does this pull-request include tests?

  - [x] Yes, it includes tests.
  - [ ] No, it does not include tests because the tests already exist.
  - [ ] No, it does not include tests because tests are too difficult to write.
  - [ ] No, it does not include tests because ...

### CODE REVIEW

File: core/array.rb [hunk 1]
```diff
@@ -1516,30 +1516,13 @@ def sample(count=undefined, options=undefined)
         while i < count
           k = rng.rand(size)
 
-          spin = false
-          spin_count = 0
-
-          while true
```
Comment (tompng) [inline core/array.rb]: Once `spin` is set to true, `spin` will never become false in this loop.
In that case, `spin_count` will always becomes over 100, and always abandoned.
(so I removed `spin`, `spin_count`, and this `while true` loop)

File: core/array.rb [hunk 2]
```diff
@@ -1516,30 +1516,13 @@ def sample(count=undefined, options=undefined)
...
-            while j < i
-              if k == result[j]
-                spin = true
-                break
-              end
-
-              j += 1
-            end
-
-            if spin
```
Comment (tompng) [inline core/array.rb]: If you add spin=false here(original algorithm with no bug?), the calculation cost might be very large.

File: core/array.rb [hunk 3]
```diff
@@ -1563,7 +1546,7 @@ def sample(count=undefined, options=undefined)
       result = Array.new self
       tuple = Rubinius::Mirror::Array.reflect(result).tuple
 
-      count.times { |i| tuple.swap i, rng.rand(size) }
+      count.times { |i| tuple.swap i, rng.rand(size - i) + i }
```
Comment (tompng) [inline core/array.rb]:
```
https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle
Implementation errors
Similarly, always selecting j from the entire range of valid array indices on every iteration also produces a result which is biased, albeit less obviously so
```

File: spec/ruby/core/array/sample_spec.rb [hunk 4]
```diff
@@ -2,6 +2,13 @@
 require File.expand_path('../fixtures/classes', __FILE__)
 
 describe "Array#sample" do
+  it "samples evenly" do
+    3.times do |i|
+      samples = 4000.times.map { 4.times.to_a.sample(3)[i] }
+      counts = samples.group_by(&:itself).map { |a, b| b.size }
+      counts.each { |count| (800..1200).should include(count) }
```
Comment (tompng) [inline spec/ruby/core/array/sample_spec.rb]: How I choose 800..1200:
probability of count in 800..1200 is 0.9999999999996083

if you change this to 900..1100, probability will be 0.99976
it means that the test wil fail in 0.024% even if there is no bug.

```ruby
def overflow_safe_mult values
  a = 1
  values = values.sort_by(&:abs)
  a *= a.abs>1 ? values.shift : values.pop until values.empty?
  a
end
def probability_of_count n, p, range
  range.map{|i|
    overflow_safe_mult [p]*i+[1-p]*(n-i)+i.times.map{|j|n-j}+i.times.map{|j|1.0/(j+1)}
  }.inject(:+)
end
probability_of_count 4000, 0.25, 0..799 #=> 3.181654463145792e-14
probability_of_count 4000, 0.25, 800..1200 #=> 0.9999999999996083
probability_of_count 4000, 0.25, 1201..4000 #=> 3.572627145186968e-13

probability_of_count 4000, 0.25, 900..1100 #=> 0.999756654675964
```

**Comment by brixen** [comment]
@tompng thank you for the PR and for the detailed explanations, much appreciated! ❤️

**Pull request merged**